Adding upload from camera action

AndyScherzinger opened the pull request:
Adds the upload from camera feature.

- [x] Add permission checks
- [x] Add processing/upload/post of the image
- [x] Add click actions/code for the camera view actions
- [x] Add retake/use + preview for confirmation
- [x] Limit image size to FHD - Resolves #1581
- [x] handle screen rotations regarding image ratio changes

|menu item|permission check (OS)|flash off (4:3,HQ)|flash on(16:9,LQ)|preview image (4:3,retake/send)|preview image (16:9,retake/send)|
|---|---|---|---|---|---|
|![device-2021-11-03-214350](https://user-images.githubusercontent.com/1315170/140190098-108d10b3-2a0b-4f72-ae94-63344028c5f9.png)|![device-2021-09-17-134654](https://user-images.githubusercontent.com/1315170/133777871-f84b998f-0a29-460e-b305-104407688e3e.png)|![device-2021-11-10-000645](https://user-images.githubusercontent.com/1315170/141020602-f5cb9dfb-610d-4e41-b810-85d8f54ed2c3.png)|![device-2021-11-10-000720](https://user-images.githubusercontent.com/1315170/141020694-64932843-e65d-46d4-ba04-0087796c2f0d.png)|![device-2021-11-10-001211](https://user-images.githubusercontent.com/1315170/141020827-8d677239-03e8-442b-9afe-ba8770109667.png)|![device-2021-11-10-000810](https://user-images.githubusercontent.com/1315170/141020851-74d811b5-73c9-49bf-86f4-4101847807d5.png)|

**Note:** Please ignore the grey bar at the top, this is due to my phone, not visible on a real device! And yes I took pictures of my monitor 😁 

Resolves #306

Signed-off-by: Andy Scherzinger <info@andy-scherzinger.de>

AndyScherzinger commented:
@mahibi ready to review ... finally 🎉

AndyScherzinger commented:
also cc @stefan-niedermann since I borrowed your code and added you as a co-author, see https://github.com/nextcloud/talk-android/commit/fdfa29b284b811078e928f2490a7c813db929ab3 - and did some changes to the UI part + previews

AndyScherzinger commented:
Ah, missed the permission checks, sry - will have to add that ...

jakobroehrl commented:
@AndyScherzinger 
Nice job!
Is it possibly to make videos directly, too?
If not, should I create an issue?

AndyScherzinger commented:
ATM it is photos only, not sure if we will support video, so, yeah a separate issue makes sense to discuss and decide on it.

nimisha-vijay commented:
It looks great! Some suggestions/questions: 
- Could we maybe change the flip camera icon to something with bigger arrows so it's more obvious? Something similar to the MDI flip camera icon: https://fonts.google.com/icons?selected=Material%20Icons%3Aflip_camera_android%3A 
I think it's okay if there aren't 2 different icons for front cam and rear cam because it's easy to figure out from the camera viewfinder itself.
- Does it make more sense to have the flip cam and flash buttons on either side of the capture button? Whatsapp and Telegram do this, but Signal does it like it is here, on the top right corner. It's not a major difference, but what do you think?
- Is there a fallback screen to if someone has not given camera permissions? A screen that says something like "Enable access to your camera to take pictures"

AndyScherzinger commented:
> Could we maybe change the flip camera icon to something with bigger arrows so it's more obvious? Something similar to the MDI flip camera icon: https://fonts.google.com/icons?selected=Material%20Icons%3Aflip_camera_android%3A
I think it's okay if there aren't 2 different icons for front cam and rear cam because it's easy to figure out from the camera viewfinder itself.

I'll replace it with the one you referenced 👍 

> Does it make more sense to have the flip cam and flash buttons on either side of the capture button? Whatsapp and Telegram do this, but Signal does it like it is here, on the top right corner. It's not a major difference, but what do you think?

Sounds good and is also how the Deck app is doing it 👍 

> Is there a fallback screen to if someone has not given camera permissions? A screen that says something like "Enable access to your camera to take pictures"

That's my last open to do. We will check if you already granted the permissions we need else a OS permission dialog will pop-up asking for the permission. 👍

AndyScherzinger commented:
@timkrueger @mahibi added the last bits and pieces, so now it is finally ready for review :)

nimisha-vijay reviewed (approved):
Looks good design wise 💯

mahibi commented:
thanks Andy! :1st_place_medal: :+1:

mahibi commented:
if i decline the permission to take photos, the popup says that "sharing" is not possible without permissions

mahibi commented:
if i take a photo in landscape view, it's displayed in portrait view and the rest of the screen is black

mahibi commented:
bug with zoom:
    - 1. switch to selfie camera
    - 2. switch to front camera
    - 3. take a photo with front camera
        --> only a zoomed part is captured

mahibi commented:
> bug with zoom:
> - 1. switch to selfie camera
> - 2. switch to front camera
> - 3. take a photo with front camera
> --> only a zoomed part is captured

if i only use front camera without switching to selfie camera it's ok:

![grafik](https://user-images.githubusercontent.com/2932790/133791782-a1ecc56b-f3cc-47f4-b96a-e23eb5c3f3aa.png)

if i use front camera after i have switched back from selfie camera:

![grafik](https://user-images.githubusercontent.com/2932790/133792070-1cadb006-8bbf-4554-b25e-1ca83dcc5dd4.png)

jancborchardt commented:
Very nice! :) Only details:
- "Upload from camera" is a bit strange wording. What do you think about "Take photo"?
- Can the "retake" button be non-primary, like the flash button in the previous view?
- Since I guess the photo will be sent to the conversation right when you press the send button, could that get a text label "Send"? :)

ASerbinski commented:
Regarding https://github.com/nextcloud/talk-android/pull/1542/commits/56f310ab89616d965e2190c5dbbf29f88945f449
I understand the need to be able to send a reduced-resolution image, maybe even by default, *however*, I think it is also important to be able to send maximum resolution very large images when it is necessary.

Can you add a setting somewhere to disable the resolution drop?

ASerbinski commented:
Target resolution "FHD" forces images to be cropped on the sides. 1920x1080 is a 16:9 ratio, however most (all?) camera sensors are 4:3.

ASerbinski commented:
I came up with this to eliminate cropping (i.e. set aspect ratio to 4:3) and set the zoom ratio to *minimum*. It also removes the fixed FHD resolution which conflicts with setting the aspect ratio. Obviously it would be nice to have some settings available for picking an available resolution (and therefore aspect ratio), and maybe a "pinch to zoom".

Note that Camera.getCameraInfo().getZoomState().getValue() provides MIN, MAX, and CURRENT zoom ratios. This would provide suitable status and limits for pinch-to-zoom or a zoom slider.

Anyway, here's my rough code;
```
diff --git a/app/build.gradle b/app/build.gradle
index 5f7148ca..1827eb2c 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -187,6 +187,7 @@ dependencies {
     ktlint "com.pinterest:ktlint:0.42.1"
     implementation 'org.conscrypt:conscrypt-android:2.5.2'
 
+    implementation 'androidx.camera:camera-core:1.0.1'
     implementation 'androidx.camera:camera-camera2:1.0.1'
     implementation 'androidx.camera:camera-lifecycle:1.0.1'
     implementation 'androidx.camera:camera-view:1.0.0-alpha28'
diff --git a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
index 9922234f..462603ae 100644
--- a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
+++ b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
@@ -30,7 +30,6 @@ import android.net.Uri;
 import android.os.Bundle;
 import android.util.DisplayMetrics;
 import android.util.Log;
-import android.util.Size;
 import android.view.OrientationEventListener;
 import android.view.Surface;
 import android.view.View;
@@ -52,6 +51,7 @@ import java.util.concurrent.ExecutionException;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.appcompat.app.AppCompatActivity;
+import androidx.camera.core.AspectRatio;
 import androidx.camera.core.Camera;
 import androidx.camera.core.ImageCapture;
 import androidx.camera.core.ImageCaptureException;
@@ -96,6 +96,7 @@ public class TakePhotoActivity extends AppCompatActivity {
                     viewModel.getCameraSelector(),
                     imageCapture,
                     preview);
+                camera.getCameraControl().setZoomRatio(camera.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
 
                 viewModel.getTorchToggleButtonImageResource()
                     .observe(
@@ -110,11 +111,12 @@ public class TakePhotoActivity extends AppCompatActivity {
                 binding.switchCamera.setOnClickListener((v) -> {
                     viewModel.toggleCameraSelector();
                     cameraProvider.unbindAll();
-                    cameraProvider.bindToLifecycle(
+                    Camera c = cameraProvider.bindToLifecycle(
                         this,
                         viewModel.getCameraSelector(),
                         imageCapture,
                         preview);
+                    c.getCameraControl().setZoomRatio(c.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
                 });
                 binding.retake.setOnClickListener((v) -> {
                     Uri uri = (Uri) binding.photoPreview.getTag();
@@ -183,7 +185,8 @@ public class TakePhotoActivity extends AppCompatActivity {
     }
 
     private ImageCapture getImageCapture() {
-        final ImageCapture imageCapture = new ImageCapture.Builder().setTargetResolution(new Size(1080, 1920)).build();
+        final ImageCapture imageCapture =
+            new ImageCapture.Builder().setTargetAspectRatio(AspectRatio.RATIO_4_3).build();
 
         orientationEventListener = new OrientationEventListener(this) {
             @Override
@@ -289,7 +292,7 @@ public class TakePhotoActivity extends AppCompatActivity {
     }
 
     private Preview getPreview() {
-        Preview preview = new Preview.Builder().build();
+        Preview preview = new Preview.Builder().setTargetAspectRatio(AspectRatio.RATIO_4_3).build();
         preview.setSurfaceProvider(binding.preview.getSurfaceProvider());
         return preview;
     }
diff --git a/app/src/main/res/layout/activity_take_picture.xml b/app/src/main/res/layout/activity_take_picture.xml
index 402583e8..ad0d4564 100644
--- a/app/src/main/res/layout/activity_take_picture.xml
+++ b/app/src/main/res/layout/activity_take_picture.xml
@@ -31,8 +31,13 @@
 
     <androidx.camera.view.PreviewView
         android:id="@+id/preview"
-        android:layout_width="match_parent"
-        android:layout_height="match_parent" />
+        android:layout_width="0dp"
+        android:layout_height="0dp"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintDimensionRatio="3:4" />
 
     <com.github.chrisbanes.photoview.PhotoView
         android:id="@+id/photo_preview"
```

ASerbinski commented:
Well actually, part of the change I made adjusted the preview window to the same aspect ratio as the preview and capture, so it all matched up. And although the display area ends up being cropped funny on a screen rotation, the captured image is correct. I'm fairly sure that I could get that all straightened out to all work out when selecting from a list of resolutions.

Zoom definitely *cannot* be handled by post-processing, since the zoom level has to be sent to the HAL, which handles the actual zoom function (by some combination of digital zoom, optical zoom, multiple sensors, and sensor configurations such as varying "pixel binning").

While post-processing crop/rotate/other may be a nice addition, its probably overkill for now.

AndyScherzinger commented:
What do you think @nickvergessen @mahibi @timkrueger ? I'd be fine with @ASerbinski's approach and we can adding anything else afterwards. I am also fine with him re-using this branch/PR.

nickvergessen commented:
I don't have enough background knowledge to judge

AndyScherzinger commented:
> I don't have enough background knowledge to judge

@nickvergessen I is just about the approach to first implement it without crop/rotate capabilities

jancborchardt commented:
@AndyScherzinger are the things from my comment adjusted? https://github.com/nextcloud/talk-android/pull/1542#issuecomment-922837367 – then I’d say it’s good to go from my side too :)

ASerbinski commented:
I have rotations working nicely now just by changing the scaleType of the PreviewView to "fitCenter" and removing the hardcoded aspect ratio from it.

```
    <androidx.camera.view.PreviewView
        android:id="@+id/preview"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        app:scaleType="fitCenter" />
```

It also looks like its not necessary to explicitly set the aspect ratio on the preview, since it picks it up from the ImageCapture.

So from here, its a matter of;
- pick a resolution from a list (the aspect ratio will come from the selected resolution),
- pinch to zoom.

This is what I have right now;
```
diff --git a/app/build.gradle b/app/build.gradle
index 5f7148ca..1827eb2c 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -187,6 +187,7 @@ dependencies {
     ktlint "com.pinterest:ktlint:0.42.1"
     implementation 'org.conscrypt:conscrypt-android:2.5.2'
 
+    implementation 'androidx.camera:camera-core:1.0.1'
     implementation 'androidx.camera:camera-camera2:1.0.1'
     implementation 'androidx.camera:camera-lifecycle:1.0.1'
     implementation 'androidx.camera:camera-view:1.0.0-alpha28'
diff --git a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
index 9922234f..7e8dc09d 100644
--- a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
+++ b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
@@ -30,7 +30,6 @@ import android.net.Uri;
 import android.os.Bundle;
 import android.util.DisplayMetrics;
 import android.util.Log;
-import android.util.Size;
 import android.view.OrientationEventListener;
 import android.view.Surface;
 import android.view.View;
@@ -52,6 +51,7 @@ import java.util.concurrent.ExecutionException;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.appcompat.app.AppCompatActivity;
+import androidx.camera.core.AspectRatio;
 import androidx.camera.core.Camera;
 import androidx.camera.core.ImageCapture;
 import androidx.camera.core.ImageCaptureException;
@@ -96,6 +96,7 @@ public class TakePhotoActivity extends AppCompatActivity {
                     viewModel.getCameraSelector(),
                     imageCapture,
                     preview);
+                camera.getCameraControl().setZoomRatio(camera.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
 
                 viewModel.getTorchToggleButtonImageResource()
                     .observe(
@@ -110,11 +111,12 @@ public class TakePhotoActivity extends AppCompatActivity {
                 binding.switchCamera.setOnClickListener((v) -> {
                     viewModel.toggleCameraSelector();
                     cameraProvider.unbindAll();
-                    cameraProvider.bindToLifecycle(
+                    Camera c = cameraProvider.bindToLifecycle(
                         this,
                         viewModel.getCameraSelector(),
                         imageCapture,
                         preview);
+                    c.getCameraControl().setZoomRatio(c.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
                 });
                 binding.retake.setOnClickListener((v) -> {
                     Uri uri = (Uri) binding.photoPreview.getTag();
@@ -183,7 +185,8 @@ public class TakePhotoActivity extends AppCompatActivity {
     }
 
     private ImageCapture getImageCapture() {
-        final ImageCapture imageCapture = new ImageCapture.Builder().setTargetResolution(new Size(1080, 1920)).build();
+        final ImageCapture imageCapture =
+            new ImageCapture.Builder().setTargetAspectRatio(AspectRatio.RATIO_4_3).build();
 
         orientationEventListener = new OrientationEventListener(this) {
             @Override
diff --git a/app/src/main/res/layout/activity_take_picture.xml b/app/src/main/res/layout/activity_take_picture.xml
index 402583e8..de83ce10 100644
--- a/app/src/main/res/layout/activity_take_picture.xml
+++ b/app/src/main/res/layout/activity_take_picture.xml
@@ -32,14 +32,15 @@
     <androidx.camera.view.PreviewView
         android:id="@+id/preview"
         android:layout_width="match_parent"
-        android:layout_height="match_parent" />
+        android:layout_height="match_parent"
+        app:scaleType="fitCenter" />
 
     <com.github.chrisbanes.photoview.PhotoView
         android:id="@+id/photo_preview"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
         android:contentDescription="@null"
-        android:scaleType="fitStart"
+        android:scaleType="fitCenter"
         android:visibility="invisible" />
 
     <com.google.android.material.button.MaterialButton
```

ASerbinski commented:
And pinch to zoom:
```
diff --git a/app/build.gradle b/app/build.gradle
index 5f7148ca..1827eb2c 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -187,6 +187,7 @@ dependencies {
     ktlint "com.pinterest:ktlint:0.42.1"
     implementation 'org.conscrypt:conscrypt-android:2.5.2'
 
+    implementation 'androidx.camera:camera-core:1.0.1'
     implementation 'androidx.camera:camera-camera2:1.0.1'
     implementation 'androidx.camera:camera-lifecycle:1.0.1'
     implementation 'androidx.camera:camera-view:1.0.0-alpha28'
diff --git a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
index 9922234f..d0b304ed 100644
--- a/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
+++ b/app/src/main/java/com/nextcloud/talk/activities/TakePhotoActivity.java
@@ -26,12 +26,14 @@ import android.content.Context;
 import android.content.Intent;
 import android.graphics.Bitmap;
 import android.graphics.BitmapFactory;
+import android.icu.number.Scale;
 import android.net.Uri;
 import android.os.Bundle;
 import android.util.DisplayMetrics;
 import android.util.Log;
-import android.util.Size;
+import android.view.MotionEvent;
 import android.view.OrientationEventListener;
+import android.view.ScaleGestureDetector;
 import android.view.Surface;
 import android.view.View;
 import android.widget.Toast;
@@ -52,11 +54,13 @@ import java.util.concurrent.ExecutionException;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.appcompat.app.AppCompatActivity;
+import androidx.camera.core.AspectRatio;
 import androidx.camera.core.Camera;
 import androidx.camera.core.ImageCapture;
 import androidx.camera.core.ImageCaptureException;
 import androidx.camera.core.Preview;
 import androidx.camera.lifecycle.ProcessCameraProvider;
+import androidx.camera.view.PreviewView;
 import androidx.core.content.ContextCompat;
 import androidx.exifinterface.media.ExifInterface;
 import androidx.lifecycle.ViewModelProvider;
@@ -76,6 +80,8 @@ public class TakePhotoActivity extends AppCompatActivity {
 
     private final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH-mm-ss", Locale.ROOT);
 
+    private Camera camera;
+
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -91,11 +97,12 @@ public class TakePhotoActivity extends AppCompatActivity {
                 final ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                 final Preview preview = getPreview();
                 final ImageCapture imageCapture = getImageCapture();
-                final Camera camera = cameraProvider.bindToLifecycle(
+                camera = cameraProvider.bindToLifecycle(
                     this,
                     viewModel.getCameraSelector(),
                     imageCapture,
                     preview);
+                camera.getCameraControl().setZoomRatio(camera.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
 
                 viewModel.getTorchToggleButtonImageResource()
                     .observe(
@@ -110,11 +117,12 @@ public class TakePhotoActivity extends AppCompatActivity {
                 binding.switchCamera.setOnClickListener((v) -> {
                     viewModel.toggleCameraSelector();
                     cameraProvider.unbindAll();
-                    cameraProvider.bindToLifecycle(
+                    camera = cameraProvider.bindToLifecycle(
                         this,
                         viewModel.getCameraSelector(),
                         imageCapture,
                         preview);
+                    camera.getCameraControl().setZoomRatio(camera.getCameraInfo().getZoomState().getValue().getMinZoomRatio());
                 });
                 binding.retake.setOnClickListener((v) -> {
                     Uri uri = (Uri) binding.photoPreview.getTag();
@@ -133,6 +141,22 @@ public class TakePhotoActivity extends AppCompatActivity {
                     finish();
                 });
 
+                ScaleGestureDetector mDetector =
+                    new ScaleGestureDetector(this, new ScaleGestureDetector.SimpleOnScaleGestureListener(){
+                        @Override
+                        public boolean onScale(ScaleGestureDetector detector){
+                            float ratio = camera.getCameraInfo().getZoomState().getValue().getZoomRatio();
+                            float delta = detector.getScaleFactor();
+                            camera.getCameraControl().setZoomRatio(ratio * delta);
+                            return true;
+                        }
+                    });
+                binding.preview.setOnTouchListener((v, event) -> {
+                    v.performClick();
+                    mDetector.onTouchEvent(event);
+                    return true;
+                });
+
                 // Enable enlarging the image more than default 3x maximumScale.
                 // Medium scale adapted to make double-tap behaviour more consistent.
                 binding.photoPreview.setMaximumScale(MAX_SCALE);
@@ -183,7 +207,8 @@ public class TakePhotoActivity extends AppCompatActivity {
     }
 
     private ImageCapture getImageCapture() {
-        final ImageCapture imageCapture = new ImageCapture.Builder().setTargetResolution(new Size(1080, 1920)).build();
+        final ImageCapture imageCapture =
+            new ImageCapture.Builder().setTargetAspectRatio(AspectRatio.RATIO_4_3).build();
 
         orientationEventListener = new OrientationEventListener(this) {
             @Override
diff --git a/app/src/main/res/layout/activity_take_picture.xml b/app/src/main/res/layout/activity_take_picture.xml
index 402583e8..de83ce10 100644
--- a/app/src/main/res/layout/activity_take_picture.xml
+++ b/app/src/main/res/layout/activity_take_picture.xml
@@ -32,14 +32,15 @@
     <androidx.camera.view.PreviewView
         android:id="@+id/preview"
         android:layout_width="match_parent"
-        android:layout_height="match_parent" />
+        android:layout_height="match_parent"
+        app:scaleType="fitCenter" />
 
     <com.github.chrisbanes.photoview.PhotoView
         android:id="@+id/photo_preview"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
         android:contentDescription="@null"
-        android:scaleType="fitStart"
+        android:scaleType="fitCenter"
         android:visibility="invisible" />
 
     <com.google.android.material.button.MaterialButton
```

nickvergessen commented:
@ASerbinski can you push commits? Or missing an invite to the Org?

ASerbinski commented:
Thank you @nickvergessen 
I've got a few hours of "day job" to get through, then I'll get on it.

AndyScherzinger commented:
@ASerbinski awesome 🎉 I'll do a rebase in a minute and then might be able to fix the design feedback from @jancborchardt 👍

AndyScherzinger commented:
@jancborchardt just did all your change requests regarding design, see updated screenshots in https://github.com/nextcloud/talk-android/pull/1542#issue-717554892

ASerbinski commented:
One thing I've never been good with is UI design, so I'm thinking of going with a very simple interface for resolution and aspect ratio.

- One toggle button to select between max and lower resolution (like 1900x[1440|1080]), 
- one toggle button to select between native and cropped (16:9) aspect ratio.

I don't think there is any pressing need to go all the way to 2:1 aspect ratio, but I suppose could be done as a 3-state toggle?

ASerbinski commented:
I currently use the same image resources as for the flash for both the crop and low resolution toggles.
I could use some help with coming up with suitable images.

These two toggles are placed at the top of the screen. Upper-left is to toggle cropping to 16x9, upper-right is to reduce resolution to 1920x1440 or 1920x1080 (depending on crop setting).

AndyScherzinger commented:
I might be able to look into the icons @ASerbinski 👍

ASerbinski commented:
I found and added some generic icons from https://fonts.google.com/icons?selected=Material+Icons

I think as far as functionality goes, it should be good now.

ASerbinski commented:
Is there anything else needed? Or is this ready?

AndyScherzinger commented:
I tried working ot the icvons over the weekend but didn't finish, sicne I guess the 2MP isn't the right icon, same for the aspect ratio, so I tried to create my own and would basically go for:

* LQ/HQ
* 16:9 / 4:3

Would that be correct @ASerbinski ? This was people choose high/low quality as well as the aspect ratio and the buttons would bevae the same way the flash behaves?

ASerbinski commented:
That could work.

AndyScherzinger commented:
@ASerbinski I changed the icons for aspect ratio and resolution in a toggling way and also moved the state into the viewmodel with live data, see latest commit. Other than that I also fixed some issues I found with the torch, like disable it when toggling front/back cam and also after a picture has been taken.

@jancborchardt can you please take a look at the screenshots again? We do have an issue in the image preview + the hovering buttons depending on the image ratio 4:3 (works fine) and 16:9 (will depend on the device). I have no (nice) idea how to improve that. So any ideas will be highly appreciated ❤️

ASerbinski commented:
I don't think I understand what you mean by there being an issue with 16x9 preview. Do you mean that the buttons go slightly over the image? I don't think I would consider that to be an issue at all. Obviously there would be some combination of device and aspect ratio that would have that happen, and I think it is actually a tasteful approach to it already.

The one thing that looks slightly odd to me is that when its in 16:9 the button says 4:3. Wouldn't it make more sense if the button reflected the state that it is currently in?

AndyScherzinger commented:
> The one thing that looks slightly odd to me is that when its in 16:9 the button says 4:3. Wouldn't it make more sense if the button reflected the state that it is currently in?

@jancborchardt @nimishavijay what do you think? @ASerbinski I did it the way flash behaves, always displaying the action that would be triggered, not the state, but maybe for this it should be different? 0_o 
Also cc @nickvergessen

jancborchardt commented:
Do we actually need the function in the first place? It should just take a photo as your normal camera would. In the same vein, is the HQ switch necessary?

Both of those do not seem like a toggle that should be directly there.

AndyScherzinger commented:
> Do we actually need the function in the first place? It should just take a photo as your normal camera would. In the same vein, is the HQ switch necessary?

There isn't much that is "necessary", but especially size has been requested to not send full-size images (since some might have a 110MP photo...).

Both of those do not seem like a toggle that should be directly there.

> So would hiding them behind a settings gear icon a solution (like my OS cam does it)?

ASerbinski commented:
@jancborchardt @AndyScherzinger ;
A settings gear icon (like where normal camera application would offer this) would be another option, but complicates implementation and makes the options less accessible. I think that both these options are very relevant and need to be easily accessed.

The HQ/LQ option really is relevant to be very easily accessible since its use would be highly correlated with an on the fly assessment of the network in use. Bad network: LQ, good network: HQ.

This application of the camera is also aiming to make a ready-to-share image almost instantly, so all other applicable options need to be easy to access. This is in contrast with the normal camera application which is more for storing the data, which can be easily run through post processing later (I.e. take an image at full native resolution, then crop and scale, then share).

jancborchardt commented:
We can also just go ahead with it as-is and see how it fares in use – then we just keep both actions.

AndyScherzinger commented:
@nickvergessen so let's merge it then? https://github.com/nextcloud/talk-android/pull/1542#issuecomment-965064030

Also pinging @mahibi and @timkrueger since @ASerbinski and I did some heavy changes after the last review.

nickvergessen commented:
Not sold on anything, I'm fine with moving forward and removing one item from the list which iOS has and Android is lagging behind.

mahibi commented:
this PR is ready for a final review @AndyScherzinger ?

AndyScherzinger commented:
@mahibi yes, the PR is complete, I did some testing and fixed everything I found. Had to do another rebase due to the latest merges to master, so CI is doing its job now. Feel free to give it a spin and review :)

timkrueger commented on app/src/main/java/com/nextcloud/talk/utils/FileUtils.java:
```diff
@@ -0,0 +1,46 @@
+package com.nextcloud.talk.utils;
```
Copyright comment is missing.

AndyScherzinger replied:
DONE :) https://github.com/nextcloud/talk-android/pull/1542/commits/4671b18b91dc4c4670438f3dd3d6689200440b32

timkrueger reviewed (approved):
Works great. Nice job. Thanks for all the work 

:1st_place_medal:

mahibi commented:
@AndyScherzinger 
screen rotation doesn't work.
once i rotate the screen, i have this behaviour:

https://user-images.githubusercontent.com/2932790/141439442-0c46cf35-d24e-4eec-8bbc-c7eb61992cde.mp4

after that it even happened often that the screen compeletly blacks out. so i had to "find" the home button (black on black) to kill the app.

mahibi commented:
great job Andy :+1: 

the only things i discovered (test with Redmi Note 7, Android 9):
- switching to landscape or portrait takes ~2 seconds while screen is black, this might be better on other devices?
- after playing around for a while the photoview was black. but i can't reproduce this right now!

so approved from my side...

AndyScherzinger commented:
Thanks @mahibi , rotation is faster for me, less than 1 sec I'd guess. Playing around I couldn't produce a black screen. So merging and let's see 👍 

Thanks for collaborating on this one @ASerbinski 🙏
